src/vidpf.rs

```diff
@@ -0,0 +1,676 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Verifiable Incremental Distributed Point Function (VIDPF).
+//!
+//! The VIDPF construction is specified in [[draft-mouris-cfrg-mastic]] and builds
+//! on techniques from [[MST23]] and [[CP22]] to lift an IDPF to a VIDPF.
+//!
+//! [CP22]: https://eprint.iacr.org/2021/580
+//! [MST23]: https://eprint.iacr.org/2023/080
+//! [draft-mouris-cfrg-mastic]: https://datatracker.ietf.org/doc/draft-mouris-cfrg-mastic/01/
+
+use core::{
+    borrow::Borrow,
+    iter::zip,
+    marker::PhantomData,
+    ops::{Add, AddAssign, Sub},
+};
+
+use rand_core::RngCore;
+use std::io::Cursor;
+use subtle::{Choice, ConditionallyNegatable, ConditionallySelectable};
+
+use crate::{
+    codec::{CodecError, Decode, Encode},
+    field::FieldElement,
+    idpf::{conditional_select_seed, conditional_xor_seeds, xor_seeds, IdpfInput, IdpfValue},
+    vdaf::xof::{Seed as XofSeed, Xof, XofFixedKeyAes128, XofTurboShake128},
+};
+
+/// Represents the domain of a point function.
+pub type VidpfInput = IdpfInput;
+
+impl VidpfInput {
+    fn iter_controlbit(&self) -> impl DoubleEndedIterator<Item = ControlBit> + '_ {
+        self.iter().rev().map(|a| ControlBit::from(a as u8))
+    }
+}
+
+/// Represents the codomain of a point function.
+pub trait VidpfValue: IdpfValue + Eq
+where
+    for<'a> &'a Self: Add<Output = Self> + Sub<Output = Self>,
+{
+}
+
+/// Creates a VIDPF instance.
+#[must_use]
+pub const fn new<V>(nonce: Vec<u8>) -> Vidpf<V>
+where
+    V: VidpfValue,
+    for<'a> &'a V: Add<Output = V> + Sub<Output = V>,
+{
+    Vidpf {
+        nonce,
+        phantom: PhantomData,
+    }
+}
+
+/// Contains the parameters of a VIDPF instance.
+pub struct Vidpf<V>
+where
+    // Represents the codomain of a point function.
+    V: VidpfValue,
+    for<'a> &'a V: Add<Output = V> + Sub<Output = V>,
+{
+    // Used to cryptographically bind some information.
+    nonce: Vec<u8>,
+
+    phantom: PhantomData<V>,
+}
+
+impl<V> Vidpf<V>
+where
+    V: VidpfValue,
+    for<'a> &'a V: Add<Output = V> + Sub<Output = V>,
+{
+    /// The [Vidpf::gen] method splits the point function `F(alpha) = beta` in two private keys
+    /// used by the aggregation servers, and a common public output share.
+    /// The `beta` value must be a non-zero value.
+    ///
+    /// The point function is defined as F: [VidpfInput] --> [VidpfValue]
+    /// such that:
+    ///
+    /// ```txt
+    /// F(x) = beta, if x == alpha.
+    /// F(x) = 0,    if x != alpha.
+    /// ```
+    pub fn gen(
+        &self,
+        alpha: &VidpfInput,
+        beta: &V,
+        parameter: &V::ValueParameter,
+    ) -> Result<(PublicShare<V>, Key, Key), VidpfError> {
+        // Key Generation.
+        let k0 = Key::gen(ServerId::S0)?;
+        let k1 = Key::gen(ServerId::S1)?;
+        self.gen_with_keys([k0, k1], alpha, beta, parameter)
+    }
+
+    /// The [Vidpf::gen_with_keys] method works as the [gen] method, except keys were already generated.
```

Comment:
nit: fixed-with font for method references, and fixing a `cargo doc --document-private-items` warning
```suggestion
    /// The [`Vidpf::gen_with_keys`] method works as the [`Vidpf::gen`] method, except keys were already generated.
```